pythonFiles/vscode_datascience_helpers/dataframes/vscodeDataFrame.py

```diff
@@ -1,11 +1,57 @@
 import pandas as _VSCODE_pd
 import builtins as _VSCODE_builtins
 import json as _VSCODE_json
+import numpy as _VSCODE_np
 import pandas.io.json as _VSCODE_pd_json
 
 # PyTorch and TensorFlow tensors which can be converted to numpy arrays
 _VSCODE_allowedTensorTypes = ["Tensor", "EagerTensor"]
 
+# Convert numpy array to DataFrame
+def _VSCODE_convertNumpyArrayToDataFrame(ndarray):
+    # Save the user's current setting
+    current_options = _VSCODE_np.get_printoptions()
+    # Ask for the full string. Without this numpy truncates to 3 leading and 3 trailing by default
+    _VSCODE_np.set_printoptions(threshold=99999)
+
+    temp = ndarray
+    x_len = temp.shape[0]
+    y_len = None
+    # Figure out if we're dealing with ragged data
+    try:
+        # Handle ragged arrays by making a container where the number of
+        # columns is the max length of all rows. Missing elements are
+        # represented as empty strings.
+        y_len = max([len(temp[i]) for i in range(x_len)])
+    except TypeError:
+        # 1D ndarray, nothing to do here
+        pass
+
+    if y_len:
+        # Figure out what kind of object the rows are
+        flattened = _VSCODE_np.full((x_len, y_len), "", dtype="object")
+        for i in range(x_len):
+            for j in range(len(temp[i])):
+                element = temp[i][j]
+                if isinstance(element, _VSCODE_np.ndarray):
+                    stringified = _VSCODE_np.array2string(element, separator=", ")
+                elif isinstance(element, (list, tuple)):
+                    # We can't pass lists and tuples to array2string because it expects
+                    # the size attribute to be defined
+                    stringified = str(element)
+                else:
+                    stringified = element
+                flattened[i][j] = stringified
+        temp = flattened
+    temp = _VSCODE_pd.DataFrame(temp)
+    ndarray = temp
+    del temp
+
+    # Restore the user's printoptions
+    _VSCODE_np.set_printoptions(threshold=current_options["threshold"])
```

Comment:
Should we possibly have everything in this function in a try and then restore this in a finally? Just worried about not restoring users print options.

Reply:
Ah you're right. I was trying to minimize my use of expansive try/catch blocks (got feedback from Eric on that in my TypeScript code 😅)

FWIW this Python code all needs to move to the Python extension. But given the hard deadline on this work for March I'm planning to get to that refactor afterwards, since that involves designing an API for providing debug info for variable views as well as the data viewer.

Reply:
Interesting, was Eric's feedback python specific or just in general? I would assume that an exception here would be a true exceptional case (i.e. not something that we know will happen from time to time) so in the non-exception thrown case I don't think this would be a perf issue.

Reply:
His feedback was general and in the context of regular TypeScript code, so I don't know that it applies here. But I wrote this particular catch block accounts with something that will happen from time to time in mind. I think there's a way to handle that error without incurring the exception, but we would still want to avoid bombing out on an unexpected error here.